File: app/src/main/java/com/shatyuka/zhiliao/hooks/PanelBubble.java
package com.shatyuka.zhiliao.hooks;

import com.shatyuka.zhiliao.Helper;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;

public class PanelBubble implements IHook {

    static List<Method> buildBubbleViewMethodList;

    @Override
    public String getName() {
        return "关闭底部气泡通知";
    }

    @Override
    public void init(ClassLoader classLoader) throws Throwable {
        Class<?> bubbleViewNew = classLoader.loadClass("com.zhihu.android.panel.ui.bubble.BubbleViewNew");
        Class<?> bubbleView = classLoader.loadClass("com.zhihu.android.panel.ui.bubble.BubbleView");

        Class<?> panelBubbleUtilNew = classLoader.loadClass("com.zhihu.android.panel.ui.bubble.n");

        buildBubbleViewMethodList = Arrays.stream(panelBubbleUtilNew.getDeclaredMethods())
                .filter(method -> method.getParameterCount() == 1)
                .filter(method -> method.getReturnType() == bubbleView
                        || method.getReturnType() == bubbleViewNew).collect(Collectors.toList());

    }

    @Override
    public void hook() throws Throwable {
        for (Method buildBubbleView : buildBubbleViewMethodList) {
            XposedBridge.hookMethod(buildBubbleView, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) {
                    if (Helper.prefs.getBoolean("switch_mainswitch", false) &&
                            Helper.prefs.getBoolean("switch_panelbubble", true)) {
                        param.setResult(null);
                    }
                }
            });
        }
    }

}

File: app/src/main/java/com/shatyuka/zhiliao/hooks/PanelBubble.kt
package com.shatyuka.zhiliao.hooks

import com.shatyuka.zhiliao.Helper
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge
import org.luckypray.dexkit.query.matchers.ClassMatcher
import java.lang.reflect.Method
import java.util.Arrays
import java.util.stream.Collectors

class PanelBubble : IHook {

    private lateinit var buildBubbleViewMethodList: List<Method>

    override fun getName(): String {
        return "关闭底部气泡通知"
    }

    @Throws(Throwable::class)
    override fun init(classLoader: ClassLoader) {
        val bubbleViewNew = classLoader.loadClass("com.zhihu.android.panel.ui.bubble.BubbleViewNew")
        val bubbleView = classLoader.loadClass("com.zhihu.android.panel.ui.bubble.BubbleView")
        val panelBubbleUtilNew = findPanelBubbleUtilNewClass(classLoader, bubbleViewNew, bubbleView)
        buildBubbleViewMethodList = Arrays.stream(panelBubbleUtilNew.getDeclaredMethods())
            .filter { method: Method -> method.parameterCount == 1 }
            .filter { method: Method ->
                (method.returnType == bubbleView || method.returnType == bubbleViewNew)
            }.collect(Collectors.toList())

        if (buildBubbleViewMethodList.isEmpty()) {
            throw NoSuchMethodException("BubbleUtilNewClass#buildBubbleView")
        }
    }

    @Throws(Throwable::class)
    override fun hook() {
        for (buildBubbleViewMethod in buildBubbleViewMethodList) {
            XposedBridge.hookMethod(buildBubbleViewMethod, object : XC_MethodHook() {
                override fun beforeHookedMethod(param: MethodHookParam) {
                    if (Helper.prefs.getBoolean("switch_mainswitch", false)
                        && Helper.prefs.getBoolean("switch_panelbubble", true)
                    ) {
                        param.setResult(null)
                    }
                }
            })
        }
    }

    @Throws(ClassNotFoundException::class)
    private fun findPanelBubbleUtilNewClass(
        classLoader: ClassLoader,
        bubbleViewNew: Class<*>,
        bubbleView: Class<*>
    ): Class<*> {

        val matcher: ClassMatcher = ClassMatcher.create().methods {
            add {
                returnType = bubbleViewNew.name
                paramCount = 1
            }
            add {
                returnType = bubbleView.name
                paramCount = 1
            }
        }

        return Helper.findClass(
            listOf("com.zhihu.android.panel.ui.bubble"), null, matcher, classLoader
        ) ?: throw ClassNotFoundException("com.zhihu.android.panel.ui.bubble.BubbleUtilNew")
    }

}